### PR TITLE
IPQualityScore integration.

### PR DESCRIPTION
**Pull request by IPQualityScore**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
Add IPQualityScore support to demisto to include support for IP, email and URL validations/verifications.

## Screenshots
https://ipqualityscore.com/downloads/screenshots/photo_2021-04-01_13-58-06.jpg
https://ipqualityscore.com/downloads/screenshots/photo_2021-04-21_13-06-29.jpg

## Minimum version of Demisto
- [x] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by dantavori** [comment]
@IPQualityScore please register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA) before we proceed with the review.

**Comment by IPQualityScore** [comment]
Hi Dan,



Thanks for the update. We've just submitted the contribution form.



We are are missing a few sections on the design docs, however all of our documentation is added to the doc.



I will see if we can supply the remaining information this week. Please let me know if any further details are needed.



Thanks,


David Mackler   


SVP Fraud Technology   


T:    | M:     


E:  ***@***.***  | W:  https://www.ipqualityscore.com/   


Skype:   David.Mackler    















---- On Sun, 25 Apr 2021 04:08:45 -0700 Dan Tavori ***@***.***> wrote ----





https://github.com/IPQualityScore please register your contribution by filling the https://forms.gle/XDfxU4E61ZwEESSMA before we proceed with the review.

File: Packs/IPQualityScore/pack_metadata.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  "email": "support@ipqualityscore.com",
+  "created": "2021-03-31T00:00:00Z",
+  "categories": [
+    "Data Enrichment & Threat Intelligence",
+	"Network Security"
+  ],
+  "tags": ["Attack", "Dark Web", "Email", "Malware", "Network", "Scam", "Security", "Security Analytics", "Spam", "VPN", "Threat Intelligence"],
+  "useCases": ["Brute Force", "Hunting", "Malware", "Phishing", "SANS", "Threat Intelligence Management", "Network Security", "Compliance", "Reputation"],
+  "keywords": ["risk score", "threat score", "botnet", "bad actor", "ATO", "takeover", "account", "cyber", "proxy", "vpn", "tor", "ip intelligence", "ip address", "ip", "domain scoring", "parked", "parking", "domain age", "abuse", "threat", "risk", "fraud", "email validation", "url scanner", "malicious url", "malicious domain", "phishing", "malware", "ip reputation", "email reputation", "domain reputation", "url reputation", "detect", "prevent", "domain", "url"]
+}
```
Comment (dantavori) [inline Packs/IPQualityScore/pack_metadata.json:17]: missing "certification" field:
```suggestion
  "keywords": ["risk score", "threat score", "botnet", "bad actor", "ATO", "takeover", "account", "cyber", "proxy", "vpn", "tor", "ip intelligence", "ip address", "ip", "domain scoring", "parked", "parking", "domain age", "abuse", "threat", "risk", "fraud", "email validation", "url scanner", "malicious url", "malicious domain", "phishing", "malware", "ip reputation", "email reputation", "domain reputation", "url reputation", "detect", "prevent", "domain", "url"],
  "certification": "certified"
}
```

File: Packs/IPQualityScore/TestPlayBooks/TestIPQualityScorePlaybook.yml [hunk 2]
```diff
@@ -0,0 +1,304 @@
+id: 261c457c-b92a-4ad6-8927-2aaa53c28ad3
```
Comment (dantavori) [inline Packs/IPQualityScore/TestPlayBooks/TestIPQualityScorePlaybook.yml:1]: please run the following demisto-sdk command to fix format issues:
```
demisto-sdk format -i Packs/IPQualityScore/TestPlayBooks/TestIPQualityScorePlaybook.yml
```
also, do we have an API key that can be used for this test playbook?

File: Packs/IPQualityScore/Integrations/IPQualityScore/IPQualityScore.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        )
+
+    def get_url_reputation(self, url: str) -> Dict[str, Any]:
+        return self._http_request(
+            method='GET',
+            url_suffix=f'{url}'
+        )
+    #
+    # def get_url_reputation(self, url: str) -> Dict[str, Any]:
+    #     pass
```
Comment (dantavori) [inline Packs/IPQualityScore/Integrations/IPQualityScore/IPQualityScore.py:35]:
```suggestion
```

File: Packs/IPQualityScore/Integrations/IPQualityScore/IPQualityScore.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    results = []
+    for ip in ips:
+        result = client.get_ip_reputation(ip)
+        result['address'] = ip
+
+        human_readable = tableToMarkdown(f"IPQualityScore Results for {ip}", result, result.keys())
+
+        if result.get('fraud_score', 0) >= ip_malicious_score_threshold:
+            score = 3
+            result['Malicious'] = {'Vendor': 'IPQualityScore'}
```
Comment (dantavori) [inline Packs/IPQualityScore/Integrations/IPQualityScore/IPQualityScore.py:74]:
```suggestion
```
no need for this line (also in the rest of the commands) - [reason](https://github.com/demisto/content/blob/master/Packs/Base/Scripts/CommonServerPython/CommonServerPython.py#L2407)

File: Packs/IPQualityScore/Integrations/IPQualityScore/IPQualityScore.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+                base_url=base_url_ip,
+                verify=verify_certificate,
+                proxy=proxy)
+            result = test_module(client)
+            demisto.results(result)
+
+        elif demisto.command() == 'ip':
+            ip_suspicious_score_threshold = int(demisto.params().get('ip_suspicious_score_threshold'))
+            ip_malicious_score_threshold = int(demisto.params().get('ip_malicious_score_threshold'))
+            reliability = demisto.params().get('feedReliability')
```
Comment (dantavori) [inline Packs/IPQualityScore/Integrations/IPQualityScore/IPQualityScore.py:218]: better define all of the parameters at the beginning of the main() function.

**Review by dantavori — CHANGES_REQUESTED** [review]
Looks great! see my (minor) notes.
My suggestions about the apikey and the URLs (in the main() function) are not mandatory, but recommended.
Please make sure that when you complete the fixes, the [validate errors](https://app.circleci.com/pipelines/github/demisto/content/81310/workflows/23e77a25-267a-4edb-8f57-752eb0b97dd7/jobs/336316) are resolved. Once you are ready, let me know and we will schedule a demo.

P.S: if you change things in the yml file, please regenerate the readme.md file to include your changes.

**Comment by IPQualityScore** [comment]
Hi Dan,



Thank you the for the revisions and recommendations.



Our team is currently analyzing your notes below. I'll provide an update for you shortly.



Please also feel welcome to test using this API Key: vzMBFeGTs9Lwj7N8AvOVil8STZKRWaTB


Thanks,


David Mackler   


SVP Fraud Technology   


T:    | M:     


E:  ***@***.***  | W:  https://www.ipqualityscore.com/   


Skype:   David.Mackler    















---- On Mon, 26 Apr 2021 05:46:20 -0700 Dan Tavori ***@***.***> wrote ----





@dantavori requested changes on this pull request.

Looks great! see my (minor) notes.
 My suggestions about the apikey and the URLs (in the main() function) are not mandatory, but recommended.
 Please make sure that when you complete the fixes, the https://app.circleci.com/pipelines/github/demisto/content/81310/workflows/23e77a25-267a-4edb-8f57-752eb0b97dd7/jobs/336316 are resolved. Once you are ready, let me know and we will schedule a demo.



In https://github.com/demisto/content/pull/12310#discussion_r620115001:

> @@ -0,0 +1,17 @@

+{

+  "name": "IPQualityScore (IPQS) Threat Risk Scoring",

+  "description": "Detect threats with real-time risk scoring by IPQS. Playbook analyzes IP addresses, email addresses, and domains or URLs for high risk behavior.",

+  "support": "partner",

+  "currentVersion": "1.0.0",

+  "author": "IPQualityScore",

+  "url": "https://www.ipqualityscore.com",

+  "email": ***@***.***",

+  "created": "2021-03-31T00:00:00Z",

+  "categories": [

+    "Data Enrichment & Threat Intelligence",

+	"Network Security"

+  ],

+  "tags": ["Attack", "Dark Web", "Email", "Malware", "Network", "Scam", "Security", "Security Analytics", "Spam", "VPN", "Threat Intelligence"],

+  "useCases": ["Brute Force", "Hunting", "Malware", "Phishing", "SANS", "Threat Intelligence Management", "Network Security", "Compliance", "Reputation"],



Reputation is not a valid use-case (causes a fail in validate), I think Threat Intelligence Management is good enough - https://github.com/michalgold please correct me if I'm wrong

⬇️ Suggested change

-  "useCases": ["Brute Force", "Hunting", "Malware", "Phishing", "SANS", "Threat Intelligence Management", "Network Security", "Compliance", "Reputation"],

+  "useCases": ["Brute Force", "Hunting", "Malware", "Phishing", "SANS", "Threat Intelligence Management", "Network Security", "Compliance"],






In https://github.com/demisto/content/pull/12310#discussion_r620115410:

> +    - contextPath: DBotScore.Score

+      type: Number

+      description: 'The actual score.'

+    - contextPath: DBotScore.Type

+      type: String

+      description: 'The indicator type.'

+    - contextPath: DBotScore.Vendor

+      type: String

+      description: 'The vendor used to calculate the score.'

+    - contextPath: URL.Data

+      description: The URL

+      type: String

+    - contextPath: URL.Malicious.Vendor

+      description: The vendor reporting the URL as malicious.

+      type: String

+  dockerimage: demisto/python3:3.9.2.17957



⬇️ Suggested change

-  dockerimage: demisto/python3:3.9.2.17957

+  dockerimage: demisto/python3:3.9.4.18682






In https://github.com/demisto/content/pull/12310#discussion_r620129514:

> @@ -0,0 +1 @@

+ 



let's add here description for obtaining the API key from IPQualityScore and configuring an integration instance




In https://github.com/demisto/content/pull/12310#discussion_r620131073:

> +    base_url_ip = urljoin('https://ipqualityscore.com/api/json/ip/', apikey)

+    base_url_email = urljoin('https://ipqualityscore.com/api/json/email/', apikey)

+    base_url_url = urljoin('https://ipqualityscore.com/api/json/url/', apikey)



better put these in the relevant client functions instead of passing them as arguments




In https://github.com/demisto/content/pull/12310#discussion_r620133702:

> +            client = Client(

+                base_url=base_url_ip,

+                verify=verify_certificate,

+                proxy=proxy)



better put this before the "if" statement to avoid code duplication (and remove the other appearances of these lines). the base_url would be https://ipqualityscore.com/api/json, and as I suggested above, the url suffix will be added in each client function.
 The api key can be passed either to the Client c'tor (best option) or as an argument to the relevant function.



In https://github.com/demisto/content/pull/12310#discussion_r620202230:

> +  display: Url Suspicious Score Threshold

+  hidden: false

+  name: url_suspicious_score_threshold

+  required: false

+  type: 0

+- additionalinfo: Threshold for fraud score from Url lookup to be marked as malicious

+  defaultvalue: '90'

+  display: Url Malicious Score Threshold

+  hidden: false

+  name: url_malicious_score_threshold

+  required: false

+  type: 0

+- additionalinfo: Reliability of the source providing the intelligence data

+  defaultvalue: F - Reliability cannot be judged

+  display: Source Reliability

+  name: feedReliability



this is not a feed integration, so the name should be:

⬇️ Suggested change

-  name: feedReliability

+  name: reliability



(don't forget to update in the .py file)




In https://github.com/demisto/content/pull/12310#discussion_r620205690:

> +def get_reputation_reliability(reliability):

+    if reliability == "A+ - 3rd party enrichment":

+        return DBotScoreReliability.A_PLUS

+    if reliability == "A - Completely reliable":

+        return DBotScoreReliability.A

+    if reliability == "B - Usually reliable":

+        return DBotScoreReliability.B

+    if reliability == "C - Fairly reliable":

+        return DBotScoreReliability.C

+    if reliability == "D - Not usually reliable":

+        return DBotScoreReliability.D

+    if reliability == "E - Unreliable":

+        return DBotScoreReliability.E

+    if reliability == "F - Reliability cannot be judged":

+        return DBotScoreReliability.F



this function is redundant - you can assign the reliability string value directly to the DBotScore. see https://github.com/demisto/content/blob/master/Packs/Base/Scripts/CommonServerPython/CommonServerPython.py#L253 and an integration https://github.com/demisto/content/blob/master/Packs/Cryptocurrency/Integrations/Cryptocurrency/Cryptocurrency.py#L32.




In https://github.com/demisto/content/pull/12310#discussion_r620210932:

> +        ip_context = Common.URL(

+            url=url,

+            dbot_score=reputation

+        )

+

+        results.append(CommandResults(

+            readable_output=human_readable,

+            indicator=ip_context,



⬇️ Suggested change

-        ip_context = Common.URL(

-            url=url,

-            dbot_score=reputation

-        )

-

-        results.append(CommandResults(

-            readable_output=human_readable,

-            indicator=ip_context,

+        url_context = Common.URL(

+            url=url,

+            dbot_score=reputation

+        )

+

+        results.append(CommandResults(

+            readable_output=human_readable,

+            indicator=url_context,






In https://github.com/demisto/content/pull/12310#discussion_r620213730:

> +        ip_context = Common.EMAIL(

+            address=email,

+            dbot_score=reputation,

+            domain=result.get('sanitized_email', email).split("@")[-1]

+        )

+

+        results.append(CommandResults(

+            readable_output=human_readable,

+            indicator=ip_context,



⬇️ Suggested change

-        ip_context = Common.EMAIL(

-            address=email,

-            dbot_score=reputation,

-            domain=result.get('sanitized_email', email).split("@")[-1]

-        )

-

-        results.append(CommandResults(

-            readable_output=human_readable,

-            indicator=ip_context,

+        email_context = Common.EMAIL(

+            address=email,

+            dbot_score=reputation,

+            domain=result.get('sanitized_email', email).split("@")[-1]

+        )

+

+        results.append(CommandResults(

+            readable_output=human_readable,

+            indicator=email_context,






In https://github.com/demisto/content/pull/12310#discussion_r620239180:

> +  "keywords": ["risk score", "threat score", "botnet", "bad actor", "ATO", "takeover", "account", "cyber", "proxy", "vpn", "tor", "ip intelligence", "ip address", "ip", "domain scoring", "parked", "parking", "domain age", "abuse", "threat", "risk", "fraud", "email validation", "url scanner", "malicious url", "malicious domain", "phishing", "malware", "ip reputation", "email reputation", "domain reputation", "url reputation", "detect", "prevent", "domain", "url"]

+}



missing "certification" field:

⬇️ Suggested change

-  "keywords": ["risk score", "threat score", "botnet", "bad actor", "ATO", "takeover", "account", "cyber", "proxy", "vpn", "tor", "ip intelligence", "ip address", "ip", "domain scoring", "parked", "parking", "domain age", "abuse", "threat", "risk", "fraud", "email validation", "url scanner", "malicious url", "malicious domain", "phishing", "malware", "ip reputation", "email reputation", "domain reputation", "url reputation", "detect", "prevent", "domain", "url"]

-}

+  "keywords": ["risk score", "threat score", "botnet", "bad actor", "ATO", "takeover", "account", "cyber", "proxy", "vpn", "tor", "ip intelligence", "ip address", "ip", "domain scoring", "parked", "parking", "domain age", "abuse", "threat", "risk", "fraud", "email validation", "url scanner", "malicious url", "malicious domain", "phishing", "malware", "ip reputation", "email reputation", "domain reputation", "url reputation", "detect", "prevent", "domain", "url"],

+  "certification": "certified"

+}






In https://github.com/demisto/content/pull/12310#discussion_r620241159:

> @@ -0,0 +1,304 @@

+id: 261c457c-b92a-4ad6-8927-2aaa53c28ad3



please run the following demisto-sdk command to fix format issues:

demisto-sdk format -i Packs/IPQualityScore/TestPlayBooks/TestIPQualityScorePlaybook.yml



also, do we have an API key that can be used for this test playbook?




In https://github.com/demisto/content/pull/12310#discussion_r620241761:

> +    #

+    # def get_url_reputation(self, url: str) -> Dict[str, Any]:

+    #     pass



⬇️ Suggested change

-    #

-    # def get_url_reputation(self, url: str) -> Dict[str, Any]:

-    #     pass






In https://github.com/demisto/content/pull/12310#discussion_r620247260:

> +    else:

+        return result.get("message", result)

+

+

+def ip_command(client, args, ip_suspicious_score_threshold, ip_malicious_score_threshold, reliability):

+    ips = argToList(args.get("ip"), ",")

+    results = []

+    for ip in ips:

+        result = client.get_ip_reputation(ip)

+        result['address'] = ip

+

+        human_readable = tableToMarkdown(f"IPQualityScore Results for {ip}", result, result.keys())

+

+        if result.get('fraud_score', 0) >= ip_malicious_score_threshold:

+            score = 3

+            result['Malicious'] = {'Vendor': 'IPQualityScore'}



⬇️ Suggested change

-            result['Malicious'] = {'Vendor': 'IPQualityScore'}



no need for this line (also in the rest of the commands) - https://github.com/demisto/content/blob/master/Packs/Base/Scripts/CommonServerPython/CommonServerPython.py#L2407




In https://github.com/demisto/content/pull/12310#discussion_r620255054:

> +            ip_suspicious_score_threshold = int(demisto.params().get('ip_suspicious_score_threshold'))

+            ip_malicious_score_threshold = int(demisto.params().get('ip_malicious_score_threshold'))

+            reliability = demisto.params().get('feedReliability')



better define all of the parameters at the beginning of the main() function.

File: Packs/IPQualityScore/pack_metadata.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  "author": "IPQualityScore",
+  "url": "https://www.ipqualityscore.com",
+  "email": "support@ipqualityscore.com",
+  "created": "2021-03-31T00:00:00Z",
+  "categories": [
+    "Data Enrichment & Threat Intelligence",
+	"Network Security"
+  ],
+  "tags": ["Attack", "Dark Web", "Email", "Malware", "Network", "Scam", "Security", "Security Analytics", "Spam", "VPN", "Threat Intelligence"],
+  "useCases": ["Brute Force", "Hunting", "Malware", "Phishing", "SANS", "Threat Intelligence Management", "Network Security", "Compliance", "Reputation"],
```
[the remark this answers is not in the document]
Comment (michalgold) [reply Packs/IPQualityScore/pack_metadata.json:15]: Yes, category and use case should cover it. We currently don't support Reputation as a use case.

**Comment by IPQualityScore** [comment]
@dantavori - Believe we're ready to go. Did make most of the changes you requested. Please let us know if you need anything else.

**Comment by IPQualityScore** [comment]
Hi Dan & Michal,



Thank you again for reviewing and your feedback.



We have pushed our latest changes which should address your recommendations below.



It looks like CircleCI may still be filling so not quite sure why that's happening.



Please feel free to test with the API key: vzMBFeGTs9Lwj7N8AvOVil8STZKRWaTB


Thanks,


David Mackler   


SVP Fraud Technology   


T:    | M:     


E:  ***@***.***  | W:  https://www.ipqualityscore.com/   


Skype:   David.Mackler    














---- On Tue, 27 Apr 2021 00:40:23 -0700 Michal Goldshtein ***@***.***> wrote ----




@michalgold commented on this pull request.


In https://github.com/demisto/content/pull/12310#discussion_r620945464:

> @@ -0,0 +1,17 @@
+{
+  "name": "IPQualityScore (IPQS) Threat Risk Scoring",
+  "description": "Detect threats with real-time risk scoring by IPQS. Playbook analyzes IP addresses, email addresses, and domains or URLs for high risk behavior.",
+  "support": "partner",
+  "currentVersion": "1.0.0",
+  "author": "IPQualityScore",
+  "url": "https://www.ipqualityscore.com",
+  "email": ***@***.***",
+  "created": "2021-03-31T00:00:00Z",
+  "categories": [
+    "Data Enrichment & Threat Intelligence",
+	"Network Security"
+  ],
+  "tags": ["Attack", "Dark Web", "Email", "Malware", "Network", "Scam", "Security", "Security Analytics", "Spam", "VPN", "Threat Intelligence"],
+  "useCases": ["Brute Force", "Hunting", "Malware", "Phishing", "SANS", "Threat Intelligence Management", "Network Security", "Compliance", "Reputation"],


Yes, category and use case should cover it. We currently don't support Reputation as a use case.

**Comment by ShahafBenYakir** [comment]
@IPQualityScore I deleted your comment as it contained credentials that are considered a secret (this is a public repo)

**Review by dantavori — APPROVED** [review]

**Pull request merged**